### PR TITLE
Refact: Notifications library and notify functions

### PR DESCRIPTION
**Pull request by kewitz**
Resolves https://github.com/opencollective/opencollective/issues/5711

Sorry about the diff, I just broke down the whole notifications lib.

### CODE REVIEW

File: server/models/Notification.ts [hunk 1]
```diff
@@ -182,6 +184,40 @@ export class Notification extends Model<InferAttributes<Notification>, InferCrea
...
 
+  static async getUnsubscribers(_where: {
+    type?: ActivityClasses | ActivityTypes;
+    CollectiveId?: number | number[];
+    UserId?: number | number[];
+  }) {
+    debug('getUnsubscribers', { _where });
+    const getUsers = notifications => notifications.map(notification => notification.User);
+
+    const include = [{ model: models.User }] as any;
```
Comment (Betree) [inline server/models/Notification.ts:195]: With:

```es6
import { Includeable } from 'sequelize';
```

```suggestion
    const include = [{ model: models.User }] as Includeable[];
```

Also curious about why we need to cast the `where` below to `any`, there's probably a more elegant solution.

Not a blocker if this is time-consuming.

File: server/models/Notification.ts [hunk 2]
```diff
@@ -182,6 +184,40 @@ export class Notification extends Model<InferAttributes<Notification>, InferCrea
...
+    where.type = flatten([where.type, `${where.type}.for.host`, ...classes]);
+
+    const unsubs = await Notification.findAll({
+      where,
+      include,
+    }).then(getUsers);
+
+    const subs = collective
+      ? await Notification.findAll({
+          where: { ...where, active: true, CollectiveId: [collective.id, collective.ParentCollectiveId] },
```
Comment (Betree) [inline server/models/Notification.ts:213]: I don't really get this, aren't unsubscribed users represented by `active: false` - in which case this function will never return them?

File: test/server/lib/recurring-contributions.test.js [hunk 3]
```diff
@@ -442,7 +388,7 @@ describe('server/lib/recurring-contributions', () => {
         expect(entry.status).to.equal('failure');
 
         // And then the nextChargeDate is ajusted for two days later
-        expect(order.Subscription.nextChargeDate.getTime()).to.equal(new Date('2018-01-30 0:0').getTime());
+        expect(order.Subscription.nextChargeDate.getTime()).to.equal(moment().startOf('day').add(2, 'days').valueOf());
```
Comment (Betree) [inline test/server/lib/recurring-contributions.test.js:391]: I don't understand why we need to update this test, seems unrelated?

**Review by Betree — COMMENTED** [review]
Great to see all this code migrated to Typescript! That's definitely a place where it adds value.

File: cron/monthly/collective-report.js [hunk 4]
```diff
@@ -210,11 +211,12 @@ const processCollective = async collective => {
...
-        type: 'collective.monthlyreport',
-        data: emailData,
-      };
-      return notifyAdminsOfCollective(collective.id, activity, options);
+      const activity = await models.Activity.create({
+        type: ActivityTypes.COLLECTIVE_MONTHLY_REPORT,
+        CollectiveId: collective.id,
+        data: { ...emailData, notify: false },
+      });
+      return notify.collective(activity, options);
```
[the remark this answers is not in the document]
Comment (kewitz) [reply cron/monthly/collective-report.js:219]: I think consistency and logging. But I agree that, since this is a cron work, it is probably not necessary.

File: server/models/Notification.ts (hunk 2 above)
Comment (kewitz) [reply server/models/Notification.ts:213]: True. The idea is being able to support exceptions at the collective level, for example:
Unsubscribe from every Transaction notification but subscribe to Transaction notifications of a specific Collective.

File: test/server/lib/recurring-contributions.test.js (hunk 3 above)
Comment (kewitz) [reply test/server/lib/recurring-contributions.test.js:391]: I had to remove the `fakeTimer` because it was getting in conflict with `waitForCondition`.

File: server/models/Notification.ts [hunk 5]
```diff
@@ -182,6 +184,47 @@ export class Notification extends Model<InferAttributes<Notification>, InferCrea
...
+    const include = [{ model: models.User }];
+    const where = { active: false };
+    if (_where.UserId) {
+      where['UserId'] = _where.UserId;
+    }
+
+    const collective = _where.CollectiveId && (await models.Collective.findByPk(_where.CollectiveId));
+    if (collective) {
+      where['CollectiveId'] = compact([collective.id, collective.ParentCollectiveId, collective.HostCollectiveId]);
+      // If user is informed, also fetch global settings
```
Comment (Betree) [inline server/models/Notification.ts:204]: I don't really understand what `informed` means in this context? Overall this entire function is pretty difficult for me to understand, it would really benefit from adding some comments to explain what we're doing at each stage.

Comment (kewitz) [reply server/models/Notification.ts:204]: Sorry about the language. The idea is that if UserId is passed as an argument, we should also consider global notification settings: basically, the setting of an UserId unsubscribe to a whole Activity Type without any CollectiveId passed to it.

**Comment by kewitz** [comment]
Since this is a big refactor, I'll be waiting until Wednesday to merge this since tomorrow I'll probably be off most of the day.

**Comment by kewitz** [comment]
Rebased and fixed conflicts.
